File: src/main/java/net/andrecarbajal/naviMusic/audio/MusicService.java
package net.andrecarbajal.naviMusic.audio;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import lombok.extern.slf4j.Slf4j;
import net.andrecarbajal.naviMusic.audio.resultHandler.AudioResultHandler;
import net.andrecarbajal.naviMusic.audio.resultHandler.SpotifyPlaylistResultHandler;
import net.andrecarbajal.naviMusic.audio.resultHandler.SpotifyTrackResultHandler;
import net.andrecarbajal.naviMusic.audio.spotify.SpotifyFetch;
import net.andrecarbajal.naviMusic.audio.spotify.SpotifyPlaylist;
import net.andrecarbajal.naviMusic.audio.spotify.SpotifySong;
import net.andrecarbajal.naviMusic.dto.response.Response;
import net.andrecarbajal.naviMusic.dto.response.RichResponse;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.managers.AudioManager;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;

@Service
@Slf4j
public class MusicService {
    private final AudioPlayerManager audioManager;
    private final SpotifyFetch spotifyFetch;

    private final Map<Long, GuildMusicManager> managers = new HashMap<>();

    public MusicService(@Lazy AudioPlayerManager audioManager, SpotifyFetch spotifyFetch) {
        this.audioManager = audioManager;
        this.spotifyFetch = spotifyFetch;
    }

    public GuildMusicManager getGuildMusicManager(Guild guild) {
        GuildMusicManager musicManager = managers.computeIfAbsent(guild.getIdLong(), a -> new GuildMusicManager(audioManager, guild));

        guild.getAudioManager().setSendingHandler(musicManager.getSendHandler());

        return musicManager;
    }

    public Response loadAndPlay(final TextChannel channel, final String provider, final String track, Member member) throws ExecutionException, InterruptedException {
        GuildMusicManager musicManager = getGuildMusicManager(channel.getGuild());

        AudioResultHandler audioResult = new AudioResultHandler(this, channel.getGuild(), member);
        audioManager.loadItemOrdered(musicManager, String.format("%s: %s", provider, track), audioResult).get();

        return audioResult.getResponse();
    }

    public Response loadAndPlayUrl(final TextChannel channel, final String trackUrl, Member member) throws ExecutionException, InterruptedException {
        GuildMusicManager musicManager = getGuildMusicManager(channel.getGuild());

        AudioResultHandler audioResult = new AudioResultHandler(this, channel.getGuild(), member);
        audioManager.loadItemOrdered(musicManager, trackUrl, audioResult).get();

        return audioResult.getResponse();
    }

    public Response loadAndPlaySpotifyUrl(TextChannel channel, final String provider, final String trackUrl, Member member) throws ExecutionException, InterruptedException {
        GuildMusicManager musicManager = getGuildMusicManager(channel.getGuild());

        if (trackUrl.toUpperCase().contains("track".toUpperCase())) {
            SpotifySong song = spotifyFetch.fetchSong(trackUrl);

            if (song == null) return new Response("Error fetching spotify song", Response.Type.ERROR, false);

            SpotifyTrackResultHandler spotifyTrackResultHandler = new SpotifyTrackResultHandler(this, channel.getGuild(), member, song);
            audioManager.loadItemOrdered(musicManager, String.format("%s: %s", provider, song), spotifyTrackResultHandler).get();

            return spotifyTrackResultHandler.getResponse();
        }

        if (trackUrl.toUpperCase().contains("playlist".toUpperCase())) {
            SpotifyPlaylist playlist = spotifyFetch.fetchPlaylist(trackUrl);
            loadSpotifySongs(playlist, channel, member, provider);
            return spotifyResponse(this, channel.getGuild(), member, playlist, "playlist");
        }

        if (trackUrl.toUpperCase().contains("album".toUpperCase())) {
            SpotifyPlaylist playlist = spotifyFetch.fetchAlbum(trackUrl);
            loadSpotifySongs(playlist, channel, member, provider);
            return spotifyResponse(this, channel.getGuild(), member, playlist, "album");
        }
        return new Response("Couldn't find spotify link", Response.Type.ERROR, false);
    }

    private Response spotifyResponse(MusicService musicService, Guild guild, Member member, SpotifyPlaylist playlist, String type) {
        RichResponse r = new RichResponse();

        r.setTitle(String.format("Adding Spotify %s to queue", type));
        r.setText(playlist.title());

        int playlistSize = playlist.songs().length;
        List<MessageEmbed.Field> fields = new ArrayList<>();
        fields.add(new MessageEmbed.Field("Songs", String.format("%d songs", playlistSize), true));

        int size = musicService.getGuildMusicManager(guild).getScheduler().getQueueSize() + playlistSize;
        fields.add(new MessageEmbed.Field("In queue", String.format(size == 1 ? "%d song" : "%d songs", size), true));

        r.setFields(fields);

        r.setFooter(new RichResponse.Footer(String.format("Added by %s", member.getEffectiveName()), member.getEffectiveAvatarUrl()));

        return r;
    }

    private void loadSpotifySongs(SpotifyPlaylist playlist, TextChannel channel, Member member, String provider) {
        for (SpotifySong song : playlist.songs()) {
            log.warn("Loading song: {}", song.toString());
            GuildMusicManager musicManager = getGuildMusicManager(channel.getGuild());
            SpotifyPlaylistResultHandler spotifyPlaylistResultHandler = new SpotifyPlaylistResultHandler(this, channel.getGuild(), member);
            audioManager.loadItemOrdered(musicManager, String.format("%s: %s", provider, song), spotifyPlaylistResultHandler);
        }
    }

    public Response skipTrack(TextChannel channel, int position) {
        GuildMusicManager musicManager = getGuildMusicManager(channel.getGuild());
        if (position < 1) return new Response("Invalid position", Response.Type.ERROR, false);

        if (position == 1) {
            musicManager.getScheduler().nextTrack();
            return new Response("Skipped", Response.Type.OK, false);
        }

        musicManager.getScheduler().skipTrack(position - 1);

        return new Response(String.format("Skipping song %d", position), Response.Type.OK, false);
    }

    public Response clear(Guild guild) {
        GuildMusicManager musicManager = getGuildMusicManager(guild);
        musicManager.getScheduler().clear();

        return new Response("Cleaning...", Response.Type.OK, false);
    }

    public void play(Guild guild, GuildMusicManager musicManager, AudioTrack track, Member member) {
        connectToChannel(guild.getAudioManager(), member);
        track.setUserData(member);
        musicManager.getScheduler().queue(track);
    }

    public void playPlaylist(Guild guild, GuildMusicManager musicManager, AudioPlaylist playlist, Member member) {
        connectToChannel(guild.getAudioManager(), member);

        for (AudioTrack track : playlist.getTracks()) {
            track.setUserData(member);
            musicManager.getScheduler().queue(track);
        }

    }

    public boolean connectToChannel(AudioManager audioManager, Member member) {
        if (audioManager.isConnected()) return false;

        try {
            audioManager.openAudioConnection(audioManager.getGuild().getVoiceChannels().stream().filter(voiceChannel -> voiceChannel.getMembers().contains(member)).findFirst().orElseThrow());
            return true;
        } catch (Exception ex) {
            log.error("Error joining voice channel", ex);
            return false;
        }

    }

    public Response nowPlaying(TextChannel textChannel) {
        GuildMusicManager musicManager = getGuildMusicManager(textChannel.getGuild());
        AudioTrack track = musicManager.getPlayer().getPlayingTrack();

        if (track == null) {
            return new Response("No track playing", Response.Type.ERROR, false);
        }

        return new Response("Now playing: " + track.getInfo().title, Response.Type.OK, false);
    }

    public Response pause(TextChannel textChannel) {
        GuildMusicManager musicManager = getGuildMusicManager(textChannel.getGuild());
        AudioPlayer player = musicManager.getPlayer();
        if (player.isPaused()) return new Response("Already paused", Response.Type.ERROR, false);

        player.setPaused(true);
        return new Response("Paused", Response.Type.OK, false);
    }

    public Response resume(TextChannel textChannel) {
        GuildMusicManager musicManager = getGuildMusicManager(textChannel.getGuild());
        AudioPlayer player = musicManager.getPlayer();
        if (!player.isPaused()) return new Response("Already playing", Response.Type.ERROR, false);

        player.setPaused(false);
        return new Response("Resumed ", Response.Type.OK, false);
    }
}

File: src/main/kotlin/net/andrecarbajal/naviMusic/audio/MusicService.kt
package net.andrecarbajal.naviMusic.audio

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import net.andrecarbajal.naviMusic.audio.resultHandler.AudioResultHandler
import net.andrecarbajal.naviMusic.audio.resultHandler.SpotifyPlaylistResultHandler
import net.andrecarbajal.naviMusic.audio.resultHandler.SpotifyTrackResultHandler
import net.andrecarbajal.naviMusic.audio.spotify.SpotifyFetch
import net.andrecarbajal.naviMusic.audio.spotify.SpotifyPlaylist
import net.andrecarbajal.naviMusic.audio.spotify.SpotifyResource
import net.andrecarbajal.naviMusic.audio.spotify.SpotifySong
import net.andrecarbajal.naviMusic.dto.response.RichResponse
import net.andrecarbajal.naviMusic.ui.MusicPresenter
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.managers.AudioManager
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Service
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import kotlin.time.Duration.Companion.milliseconds

@Service
class MusicService(
    @Lazy private val audioPlayerManager: AudioPlayerManager,
    private val spotifyFetch: SpotifyFetch,
    private val presenter: MusicPresenter,
    @Value($$"${app.music.max-pending-loads:20}") private val maxPendingLoads: Int,
    @Value($$"${app.music.max-queue-size:500}") private val maxQueueSize: Int,
    @Value($$"${app.music.max-playlist-tracks:100}") private val maxPlaylistTracks: Int
) {
    private val log = LoggerFactory.getLogger(MusicService::class.java)
    private val managers = ConcurrentHashMap<Long, GuildMusicManager>()
    private val serviceScope = CoroutineScope(Dispatchers.IO)

    fun getGuildMusicManager(guild: Guild): GuildMusicManager = managers.computeIfAbsent(guild.idLong) {
        GuildMusicManager(audioPlayerManager, guild, maxPendingLoads) { releaseGuild(guild) }
    }.also { guild.audioManager.sendingHandler = it.getSendHandler() }

    fun releaseGuild(guild: Guild) {
        managers.remove(guild.idLong)?.shutdown()
    }

    fun loadAndPlay(
        channel: TextChannel,
        provider: String,
        track: String,
        member: Member?,
        event: SlashCommandInteractionEvent? = null
    ): RichResponse {
        val accepted = loadItem(channel.guild, member, "$provider: $track") { done ->
            AudioResultHandler(this, channel.guild, member!!, event, null, done)
        }
        return loadingResponse(accepted, "Loading...", "Adding track to queue")
    }

    fun loadAndPlayUrl(
        channel: TextChannel,
        trackUrl: String,
        member: Member?,
        event: SlashCommandInteractionEvent? = null
    ): RichResponse {
        val accepted = loadItem(channel.guild, member, trackUrl) { done ->
            AudioResultHandler(this, channel.guild, member!!, event, trackUrl, done)
        }
        return loadingResponse(accepted, "Loading URL...", "Fetching data from provided link")
    }

    fun loadAndPlaySpotifyUrl(
        channel: TextChannel,
        provider: String,
        trackUrl: String,
        member: Member?,
        event: SlashCommandInteractionEvent? = null
    ): RichResponse {
        if (!spotifyFetch.isConfigured()) return presenter.formatSimpleResponse(
            "Spotify No Configurado",
            "El bot no tiene configurada la API de Spotify.",
            RichResponse.Type.USER_ERROR
        )
        val manager = getGuildMusicManager(channel.guild)
        val accepted = manager.enqueueLoad { done ->
            serviceScope.launch {
                try {
                    when (val resource = SpotifyResource.fromUrl(trackUrl)) {
                        is SpotifyResource.Track -> spotifyFetch.fetchSong(resource.id)?.let { song ->
                            loadSpotifySong(channel.guild, member, provider, song, event, done)
                        } ?: done()

                        is SpotifyResource.Playlist, is SpotifyResource.Album -> {
                            val playlist =
                                if (resource is SpotifyResource.Playlist) spotifyFetch.fetchPlaylist(resource.id) else spotifyFetch.fetchAlbum(
                                    (resource as SpotifyResource.Album).id
                                )
                            if (playlist == null) done() else loadSpotifyPlaylist(
                                manager,
                                channel,
                                member,
                                provider,
                                playlist,
                                resource,
                                trackUrl,
                                event,
                                done
                            )
                        }

                        is SpotifyResource.Invalid -> done()
                    }
                } catch (exception: Exception) {
                    log.error("Error processing Spotify URL", exception)
                    done()
                }
            }
        }
        return loadingResponse(accepted, "Spotify Processing", "Analyzing Spotify link...")
    }

    private fun loadSpotifySong(
        guild: Guild,
        member: Member?,
        provider: String,
        song: SpotifySong,
        event: SlashCommandInteractionEvent?,
        done: () -> Unit
    ) {
        connectToChannel(guild.audioManager, member).thenAccept { connected ->
            if (!connected) done() else audioPlayerManager.loadItem(
                "$provider: $song",
                SpotifyTrackResultHandler(this, guild, member, song, event, done)
            )
        }
    }

    private fun loadSpotifyPlaylist(
        manager: GuildMusicManager,
        channel: TextChannel,
        member: Member?,
        provider: String,
        playlist: SpotifyPlaylist,
        resource: SpotifyResource,
        trackUrl: String,
        event: SlashCommandInteractionEvent?,
        done: () -> Unit
    ) {
        val songs = playlist.songs.take(maxPlaylistTracks)
        if (playlist.songs.size > maxPlaylistTracks) log.warn(
            "Spotify collection {} truncated to {} tracks",
            playlist.title,
            maxPlaylistTracks
        )
        connectToChannel(channel.guild.audioManager, member).thenAccept { connected ->
            if (!connected || songs.isEmpty()) {
                done(); return@thenAccept
            }
            event?.let {
                val type = if (resource is SpotifyResource.Playlist) "playlist" else "album"
                manager.submit { manager.scheduler.getQueueSize() }.thenAccept { size ->
                    presenter.formatSpotifyResponse(member, playlist, type, trackUrl, size).editReply(it)
                }
            }
            loadSpotifySongsSequentially(manager, channel.guild, member, provider, songs, 0, done)
        }
    }

    private fun loadSpotifySongsSequentially(
        manager: GuildMusicManager,
        guild: Guild,
        member: Member?,
        provider: String,
        songs: List<SpotifySong>,
        index: Int,
        done: () -> Unit
    ) {
        if (index >= songs.size) {
            done(); return
        }
        audioPlayerManager.loadItem("$provider: ${songs[index]}", SpotifyPlaylistResultHandler(this, guild, member) {
            loadSpotifySongsSequentially(manager, guild, member, provider, songs, index + 1, done)
        })
    }

    private fun loadItem(
        guild: Guild,
        member: Member?,
        identifier: String,
        handler: (() -> Unit) -> AudioLoadResultHandler
    ): Boolean {
        val manager = getGuildMusicManager(guild)
        return manager.enqueueLoad { done ->
            connectToChannel(guild.audioManager, member).thenAccept { connected ->
                if (connected) audioPlayerManager.loadItem(identifier, handler(done)) else done()
            }
        }
    }

    private fun loadingResponse(accepted: Boolean, title: String, text: String): RichResponse =
        if (accepted) presenter.formatSimpleResponse(title, text)
        else presenter.formatSimpleResponse(
            "Queue busy",
            "Too many pending requests in this server. Try again shortly.",
            RichResponse.Type.USER_ERROR
        )

    fun play(musicManager: GuildMusicManager, track: AudioTrack, member: Member?): CompletableFuture<Unit> =
        musicManager.submit {
            require(musicManager.scheduler.getQueueSize() < maxQueueSize) { "Queue limit reached" }
            track.userData = member
            musicManager.scheduler.queue(track)
        }

    fun playPlaylist(
        musicManager: GuildMusicManager,
        playlist: AudioPlaylist,
        member: Member?
    ): CompletableFuture<Unit> = musicManager.submit {
        val tracks = playlist.tracks.take(maxPlaylistTracks)
        require(musicManager.scheduler.getQueueSize() + tracks.size <= maxQueueSize) { "Queue limit reached" }
        tracks.forEach { it.userData = member; musicManager.scheduler.queue(it) }
    }

    fun skipTrack(channel: TextChannel, position: Int): RichResponse = getGuildMusicManager(channel.guild).submit {
        val scheduler = getGuildMusicManager(channel.guild).scheduler
        if (position < 1) return@submit presenter.formatSimpleResponse(
            "Error",
            "Invalid position",
            RichResponse.Type.USER_ERROR
        )
        val tracks = scheduler.queue.toList()
        val track =
            if (position == 1) getGuildMusicManager(channel.guild).player.playingTrack else tracks.getOrNull(position - 2)
        if (track == null) return@submit presenter.formatSimpleResponse(
            "Error",
            "Position $position doesn't exist",
            RichResponse.Type.USER_ERROR
        )
        if (position == 1) scheduler.nextTrack() else scheduler.skipTrack(position - 1)
        presenter.formatSimpleResponse("Skipping...", "Removed: [${track.info.title}](${track.info.uri})")
    }.join()

    fun clear(guild: Guild): RichResponse = getGuildMusicManager(guild).submit {
        getGuildMusicManager(guild).scheduler.clear()
        presenter.formatSimpleResponse("Queue cleaned", "All songs removed.")
    }.join()

    fun leave(guild: Guild) {
        val manager = managers[guild.idLong]
        manager?.submit { manager.scheduler.clear(); manager.player.stopTrack(); guild.audioManager.closeAudioConnection() }
            ?.join()
        releaseGuild(guild)
    }

    fun shuffle(guild: Guild) {
        getGuildMusicManager(guild).submit { getGuildMusicManager(guild).scheduler.shuffle() }
    }

    fun setRepeating(guild: Guild, repeating: Boolean): Boolean {
        val manager = getGuildMusicManager(guild)
        return manager.submit {
            val old = manager.scheduler.isRepeating; manager.scheduler.isRepeating = repeating; old
        }.join()
    }

    fun isRepeating(guild: Guild): Boolean {
        val manager = getGuildMusicManager(guild)
        return manager.submit { manager.scheduler.isRepeating }.join()
    }

    fun queueSnapshot(guild: Guild): Pair<AudioTrack?, List<AudioTrack>> {
        val manager = getGuildMusicManager(guild)
        return manager.submit { manager.player.playingTrack to manager.scheduler.queue.toList() }.join()
    }

    fun connectToChannel(audioManager: AudioManager, member: Member?): CompletableFuture<Boolean> {
        if (audioManager.isConnected) return CompletableFuture.completedFuture(true)
        val voiceChannel = member?.voiceState?.channel ?: return CompletableFuture.completedFuture(false)
        val future = CompletableFuture<Boolean>()
        try {
            audioManager.openAudioConnection(voiceChannel)
            serviceScope.launch {
                repeat(100) {
                    if (audioManager.isConnected) {
                        delay(2000.milliseconds); future.complete(true); return@launch
                    }; delay(100.milliseconds)
                }
                future.complete(false)
            }
        } catch (exception: Exception) {
            log.error("Error joining voice channel", exception); future.complete(false)
        }
        return future
    }

    fun nowPlaying(textChannel: TextChannel): RichResponse {
        val manager = getGuildMusicManager(textChannel.guild)
        return manager.submit {
            val track = manager.player.playingTrack ?: return@submit presenter.formatSimpleResponse(
                "No music",
                "The queue is empty.",
                RichResponse.Type.ERROR
            )
            presenter.formatNowPlaying(track)
        }.join()
    }

    fun pause(textChannel: TextChannel): RichResponse {
        val manager = getGuildMusicManager(textChannel.guild)
        return manager.submit {
            if (manager.player.isPaused) presenter.formatSimpleResponse(
                "Error",
                "Already paused",
                RichResponse.Type.USER_ERROR
            ) else {
                manager.player.isPaused = true; presenter.formatSimpleResponse("Paused", "Use /resume to continue")
            }
        }.join()
    }

    fun resume(textChannel: TextChannel): RichResponse {
        val manager = getGuildMusicManager(textChannel.guild)
        return manager.submit {
            if (!manager.player.isPaused) presenter.formatSimpleResponse(
                "Error",
                "Already playing",
                RichResponse.Type.USER_ERROR
            ) else {
                manager.player.isPaused = false; presenter.formatSimpleResponse("Resumed", "Playback resumed")
            }
        }.join()
    }
}
